test/xbgpu/test_beamform.py

```diff
@@ -0,0 +1,114 @@
+################################################################################
+# Copyright (c) 2023, National Research Foundation (SARAO)
+#
+# Licensed under the BSD 3-Clause License (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Test the :mod:`katgpucbf.xbgpu.beamform` module."""
+
+import numba
+import numpy as np
+import pytest
+from katsdpsigproc.abc import AbstractCommandQueue, AbstractContext
+
+from katgpucbf import COMPLEX, N_POLS
+from katgpucbf.xbgpu.beamform import BeamformTemplate
+
+
+@numba.njit
+def quant(x):
+    """Round to integer and clamp to [-127, 127]."""
+    return np.fmin(np.fmax(np.round(x), -127), 127)
+
+
+@numba.njit
+def beamform_host(in_: np.ndarray, weights: np.ndarray, delays: np.ndarray, beam_pols: np.ndarray) -> np.ndarray:
+    """Implement the beamforming operation on the CPU.
+
+    The input arrays have the same shape and meaning as in :mod:`katgpucbf.xbgpu.beamform`.
+    """
+    n_frames = in_.shape[0]
+    n_channels = in_.shape[2]
+    n_times = in_.shape[3]
+    n_beams = len(beam_pols)
+    out = np.zeros((n_frames, n_beams, n_channels, n_times, COMPLEX), np.int8)
+    for frame in range(n_frames):
+        for channel in range(n_channels):
+            in_c = in_[frame, :, channel, ..., 0] + np.complex64(1j) * in_[frame, :, channel, ..., 1]
+            for beam in range(n_beams):
+                p = beam_pols[beam]
+                w = weights[:, beam]
+                w = w * np.exp(delays[:, beam] * channel * np.complex64(1j) * np.float32(np.pi))
+                for time in range(n_times):
+                    # .copy() is added because otherwise we get a warning that np.dot is
+                    # more efficient on C-contiguous arrays.
+                    accum = np.dot(in_c[:, time, p].copy(), w)
+                    out[frame, beam, channel, time, 0] = quant(accum.real)
+                    out[frame, beam, channel, time, 1] = quant(accum.imag)
+    return out
```

Comment:
> Basically, I don't see an antenna dimension being looped over, just asking what's up with that

It's hidden inside the `np.dot`.

> Lastly, your choice of naming convention for in_, with the trailing underscore

It's because `in` is a keyword, so I can't use that as the variable name.

Reply:
`in_c[:, time, p]` selects a slice out of a multi-dimensional array (without copying it), and the elements it selects are not contiguous in memory.